### DISCUSSION TITLE
question: why directed graph ?

### DISCUSSION THREAD

**Discussion by clojj** [srid/neuron#53] [General]
Great project!

why only directed graphs ?
I imagine zettels to be linked completely free in any way possible.
Is this because of Zettelkasten philosophy or because of the current graph tooling in neuron ?

**Comment by srid** [comment]
You can always derive an undirected graph from a directed graph (by adding a flipped copy of each edge). So strictly speaking neuron provides both directed and undirected graph. And yes, zettels can be linked completely free; a link between two zettels is no different to a link between two webpages on the internet.

Having a direction is very useful in visualizations like [automatic category tree](https://neuron.srid.ca/2011503.html), and knowing the 'backlinks' of a zettel.

But, in the future, we can use the undirected graph to render a mind map view of the zettelkasten.

So, I suppose, which graph we use is a matter of how we want to visualize it:

- directed graph -> category tree, or category forest
- undirected graph -> mind map, or [network graph](https://www.d3-graph-gallery.com/network)

Thoughts?

**Comment by clojj** [comment]
That makes sense, thx
Visualization is a separate concern, yes.

Also, I just discovered, that right now, neuron stores everything in files (and reads all of them initially ?)
This may well be Ok for a reasonable sized zettel-collection, but did you investigate something like a graph-database yet ? or some other NoSQL store with history(!) and full-text search ?
Just wondering...

**Comment by srid** [comment]
Neuron will always be based on plain-text files. 

I'm not worried about performance. If/when the Zettelkasten grows to be huge and actually becomes a bottleneck, we would be in a better position to address the specific performance issues in various ways, such as by automatically creating an index of zettel notes.

From the user point of view, their data will continue to remain nothing more than a bunch of Markdown files.

**Comment by srid** [comment]
Moving this [discussion](https://funprog.zulipchat.com/#narrow/stream/231929-Neuron/topic/Plain.20text) to [Zulip](https://funprog.zulipchat.com/#narrow/stream/231929-Neuron/topic/Undirected.20graph)
